packages/vscode-extension/src/debug/teamsfxTaskHandler.ts

```diff
@@ -376,17 +381,45 @@ function onDidTerminateDebugSessionHandler(event: vscode.DebugSession): void {
     }
 
     allRunningDebugSessions.delete(event.id);
+    if (allRunningDebugSessions.size == 0) {
+      endLocalDebugSession();
+    }
     allRunningTeamsfxTasks.clear();
   }
 }
 
 export function registerTeamsfxTaskAndDebugEvents(): void {
-  ext.context.subscriptions.push(vscode.tasks.onDidStartTaskProcess(onDidStartTaskProcessHandler));
-  ext.context.subscriptions.push(vscode.tasks.onDidEndTaskProcess(onDidEndTaskProcessHandler));
   ext.context.subscriptions.push(
-    vscode.debug.onDidStartDebugSession(onDidStartDebugSessionHandler)
+    vscode.tasks.onDidStartTaskProcess((event: vscode.TaskProcessStartEvent) =>
+      Correlator.runWithId(getLocalDebugSessionId(), onDidStartTaskProcessHandler, event)
+    )
+  );
+
+  ext.context.subscriptions.push(
+    vscode.tasks.onDidEndTaskProcess((event: vscode.TaskProcessEndEvent) =>
+      Correlator.runWithId(getLocalDebugSessionId(), onDidEndTaskProcessHandler, event)
+    )
+  );
+
+  // debug session handler use correlation-id from event.configuration.teamsfxCorrelationId
+  // to minimize concurrent debug session affecting correlation-id
+  ext.context.subscriptions.push(
+    vscode.debug.onDidStartDebugSession((event: vscode.DebugSession) =>
+      Correlator.runWithId(
+        // fallback to retrieve correlation id from the global variable.
+        event.configuration.teamsfxCorrelationId || getLocalDebugSessionId(),
+        onDidStartDebugSessionHandler,
+        event
+      )
+    )
   );
   ext.context.subscriptions.push(
-    vscode.debug.onDidTerminateDebugSession(onDidTerminateDebugSessionHandler)
+    vscode.debug.onDidTerminateDebugSession((event: vscode.DebugSession) =>
+      Correlator.runWithId(
+        event.configuration.teamsfxCorrelationId || getLocalDebugSessionId(),
```

Comment:
Why not always `event.configuration.teamsfxCorrelationId` or `getLocalDebugSessionId()`. What's the mismatch case?

Reply:
1. Prefer using `event.configuration.teamsfxCorrelationId` because `event` is associate with one vscode debug session, so concurrent f5 will not cause incorrect correlation id after this.
2. fallback to `getLocalDebugSessionId()` because the backend debug will not call `TeamsfxDebugHandler.resolveDebugConfiguration()` so `event.configuration.teamsfxCorrelationId` is undefined.